### ISSUE TITLE
Detect whether an output symbolically depends on an input

### ISSUE THREAD

**Issue by danijar** [google/jax#10867]
Currently, `jax.grad()` returns zero tensors for gradients where the output does not depend on the input. However, this typically scenario indicates a user error that can be difficult to detect and should not be silently ignored. TensorFlow returns `None` for gradients of outputs with respect to inputs when there is no symbolic path (without stop_gradient on the way) between the input and the output. Something like this would be useful to add to JAX to catch subtle bugs in user code.

For example, there could be an additional keyword argument `jax.grad(..., indep_as_none=False)` and setting it to `True` would return a pytree with `None` for inputs that have no path to the output in the compute graph. This would make it easy for users to check whether there are any `None` values in the gradient that likely indicate an error.

```python3
def fn(x):
  return 2 * x, jnp.zeros(())

x = jnp.ones(())

g = jax.grad(fn)(x)
print(g)  # (2, 0)

g = jax.grad(fn, indep_as_none=True)(x)
print(g)  # (2, None)
```

**Comment by mattjj** [comment]
Thanks for the clear feature request! This does sound useful, though we probably wouldn't build it into `grad` by default.

Here's a way to do it **using JAX-internal APIs which might break without warning in the future**:

```python
import jax
from jax import linear_util as lu
from jax.interpreters import ad
from jax._src import ad_util
from jax._src.api_util import flatten_fun
from jax.tree_util import tree_flatten, tree_unflatten

def grad_type(fun, *args, **kwargs):
  in_primals, in_tree = tree_flatten((args, kwargs))
  flat_fun, _ = flatten_fun(lu.wrap_init(fun), in_tree)
  out_primals, out_pvals, jaxpr, consts = ad.linearize(flat_fun, *in_primals)
  dummy_args = [ad.UndefinedPrimal(v.aval) for v in jaxpr.invars]
  cts = [ct for ct, pval in zip(out_primals, out_pvals) if not pval.is_known()]
  nonzero_cotangents = None
  def backward_pass(cts):
    nonlocal nonzero_cotangents
    arg_cts = ad.backward_pass(jaxpr, (), False, consts, dummy_args, cts)
    nonzero_cotangents = [type(ct) is not ad_util.Zero for ct in arg_cts]
  dummy_cts = [y for y, pval in zip(out_primals, out_pvals) if not pval.is_known()]
  jax.make_jaxpr(backward_pass)(dummy_cts)
  return tree_unflatten(in_tree, nonzero_cotangents)

print(grad_type(lambda x, y: 2 * x, 3., 4.))
# ((True, False), {})


def f(x):
  if x['hi'] > 0:
    return x['hi'] * 5.
  else:
    return x['bye'] * 6

print(grad_type(f, {'hi': 3., 'bye': 4}))
# (({'bye': False, 'hi': True},), {})
```

**Comment by danijar** [comment]
Having it as a separate function sounds like a good idea, as long as it can still be used to catch errors at JIT compile time without slowing down training time of the model later on?

**Comment by mattjj** [comment]
That's right, it won't slow down any `jit`ted function executions.
